File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    double myPow(double x, int n) {
        long long z = n;
        if (z< 0) {
        x = 1 / x;
        z = -z;
        }
        if (z == 0) return 1;
            double half = myPow(x, z / 2);
        return (z% 2) ? x * half * half: half * half;
    }
};

File: 0050-powx-n/0050-powx-n.py
class Solution(object):
    def myPow(self, x, n):
        ans = 1.0
        nn = n
        if n == 0:
            return 1.0
        if n<0:
            return 1/ self.myPow(x,-n)
        half = self.myPow(x , n//2)
        if n%2==0:
            return half * half
        else:
            return half * half  * x